inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp

```diff
@@ -427,25 +429,33 @@ void kernels_cache::build_all() {
     if (!_pending_compilation)
         return;
 
-    std::lock_guard<std::mutex> lock(_context.get_cache_mutex());
-
-    auto sorted_program_code = get_program_source(_kernels_code);
+    kernels_cache::sorted_code sorted_program_code;
+    {
+        std::lock_guard<std::mutex> lock(_context.get_cache_mutex());
+        sorted_program_code = get_program_source(_kernels_code);
+        _one_time_kernels.clear();
+    }
 
-    _one_time_kernels.clear();
+    std::vector<std::future<void>> builds;
     for (auto& program : sorted_program_code) {
-        auto kernels = build_program(program.second);
-
-        for (auto& k : kernels) {
-            const auto& entry_point = k.first;
-            const auto& k_id = program.second.entry_point_to_id[entry_point];
-            if (program.second.one_time) {
-                _one_time_kernels[k_id] = k.second;
-            } else {
-                _kernels[k_id] = k.second;
+        builds.push_back(std::async(std::launch::async,[&]() {
+            auto kernels = build_program(program.second);
+
+            for (auto &k : kernels) {
```

Comment:
BTW, we have a parameter that controls how many kernels are grouped into single compilation unit (`get_max_kernels_per_batch()` method returns this value. Currently it's equal to `10`). Maybe it makes sense to lower this value to have more efficient parallelism?

Reply:
I will do some experiments with that

Reply:
Really, I don't quite understand the initial purpose of the get_max_kernels_per_batch()- it results in creating additional sub-programs, that are not easy  to compile in parallel (at least with my original approach).
Instead, I have added a snippet that creates at least 2 OCL programs, and this better accelerates the topologies I tried (.e.g. the BERT is now loaded 1.5x than my initial code, and almost 2x compared to the master).

Yet, overall, I feel like I'm less familiar with your code, than you of course, so pls carefully review my PR, as I really would want to avoid any potential issues. Also, per @yury-gorbachev, next time I should rather create a feature request  for the GPU team, rather than do my own impl. So feel free to close the PR and redo the feature yourself, if you still have opens.